modules/launcher/src/main/java/org/mule/runtime/module/launcher/application/DefaultApplicationFactory.java

```diff
@@ -129,28 +145,45 @@ private ArtifactClassLoader createCompositePluginClassLoader(ArtifactClassLoader
         return new CompositeArtifactClassLoader("appPlugins", parent.getClassLoader(), classLoaders, parent.getClassLoaderLookupPolicy());
     }
 
-    private List<ApplicationPlugin> createApplicationPlugins(ArtifactClassLoader parentClassLoader, Set<ApplicationPluginDescriptor> pluginDescriptors)
+    private List<ApplicationPlugin> createContainerApplicationPlugins(ArtifactClassLoader parentClassLoader, ApplicationDescriptor appDescriptor) throws IOException
     {
         final List<ApplicationPlugin> plugins = new LinkedList<>();
 
-        for (ApplicationPluginDescriptor descriptor : pluginDescriptors)
+        File[] containerPlugins = MuleFoldersUtil.getContainerAppPluginsFolder().listFiles();
+        if (containerPlugins != null)
         {
-            final MuleArtifactClassLoader pluginClassLoader = createPluginClassLoader(parentClassLoader, descriptor);
-            final DefaultApplicationPlugin applicationPlugin = new DefaultApplicationPlugin(descriptor, pluginClassLoader);
-
-            plugins.add(applicationPlugin);
+            for (File pluginFolder : MuleFoldersUtil.getContainerAppPluginsFolder().listFiles((FileFilter) new SuffixFileFilter(".zip", IOCase.INSENSITIVE)))
+            {
+                String pluginName = FilenameUtils.removeExtension(pluginFolder.getName());
+                if (appDescriptor.containsApplicationPlugin(pluginName))
```

Comment:
I would separate this check from the container plugin creation logic. ie, first create the container plugins, then check if there are duplicates. Deployment could take more time to fail, but the code would be a bit more clear